Snappy codec does not work

When writing Avro data with the Snappy codec, it cannot be read by the Java implementation from Apache. And vice versa.

To reproduce, run this Rust program and redirect stdout to a file (`snappy.avro`):
```
[dependencies.avro-rs]
version = "0.6.1"
features = ["snappy"]
```

```
extern crate avro_rs;

use avro_rs::Codec;
use avro_rs::Schema;
use avro_rs::types::Record;
use avro_rs::Writer;
use std::io::{self, Write};

fn main() {
    let raw_schema = r#"
        {
            "type": "record",
            "name": "test",
            "fields": [
                {"name": "a", "type": "long", "default": 42},
                {"name": "b", "type": "string"}
            ]
        }
    "#;

    let schema = Schema::parse_str(raw_schema).unwrap();

    let mut writer = Writer::with_codec(&schema, io::stdout(), Codec::Snappy);
    let mut record = Record::new(writer.schema()).unwrap();
    record.put("a", 27i64);
    record.put("b", "foo");
    writer.append(record).unwrap();
    writer.flush().unwrap();
}

```

Then run
```
java -jar avro-tools-1.8.2.jar tojson snappy.avro
```

I get this error:
```
Exception in thread "main" org.apache.avro.AvroRuntimeException: java.io.IOException: FAILED_TO_UNCOMPRESS(5)
        at org.apache.avro.file.DataFileStream.hasNext(DataFileStream.java:210)
        at org.apache.avro.tool.DataFileReadTool.run(DataFileReadTool.java:76)
        at org.apache.avro.tool.Main.run(Main.java:87)
        at org.apache.avro.tool.Main.main(Main.java:76)
Caused by: java.io.IOException: FAILED_TO_UNCOMPRESS(5)
        at org.xerial.snappy.SnappyNative.throw_error(SnappyNative.java:84)
        at org.xerial.snappy.SnappyNative.rawUncompress(Native Method)
        at org.xerial.snappy.Snappy.rawUncompress(Snappy.java:444)
        at org.xerial.snappy.Snappy.uncompress(Snappy.java:480)
        at org.apache.avro.file.SnappyCodec.decompress(SnappyCodec.java:60)
        at org.apache.avro.file.DataFileStream$DataBlock.decompressUsing(DataFileStream.java:355)
        at org.apache.avro.file.DataFileStream.hasNext(DataFileStream.java:199)
        ... 3 more

```

(If I use Null or Deflate codec, it works as expected.)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## [Unreleased]
+### Fixed
+- Snappy codec
 
 ## [0.6.1]- 2018-10-07
 ### Fixed

--- a/Cargo.toml
+++ b/Cargo.toml
@@ -8,9 +8,11 @@ readme = "README.md"
 repository = "https://github.com/flavray/avro-rs"
 
 [features]
-snappy = ["snap"]
+snappy = ["byteorder", "crc", "snap"]
 
 [dependencies]
+byteorder = { version = "1.0.0", optional = true }
+crc = { version = "1.3.0", optional = true }
 failure = "0.1.1"
 failure_derive = "0.1.1"
 libflate = "0.1"

--- a/src/codec.rs
+++ b/src/codec.rs
@@ -5,7 +5,9 @@ use std::str::FromStr;
 use failure::Error;
 use libflate::deflate::{Decoder, Encoder};
 #[cfg(feature = "snappy")]
-use snap::{Reader, Writer};
+use byteorder;
+#[cfg(feature = "snappy")]
+use crc;
 
 use types::{ToAvro, Value};
 use util::DecodeError;
@@ -66,9 +68,16 @@ impl Codec {
             },
             #[cfg(feature = "snappy")]
             Codec::Snappy => {
-                let mut writer = Writer::new(Vec::new());
-                writer.write_all(stream)?;
-                *stream = writer.into_inner()?; // .into_inner() will also call .flush()
+                use byteorder::ByteOrder;
+
+                let mut encoded: Vec<u8> = vec!(0; snap::max_compress_len(stream.len()));
+                let compressed_size = snap::Encoder::new().compress(&stream[..], &mut encoded[..])?;
+
+                let crc = crc::crc32::checksum_ieee(&stream[..]);
+                byteorder::BigEndian::write_u32(&mut encoded[compressed_size..], crc);
+                encoded.truncate(compressed_size + 4);
+
+                *stream = encoded;
             },
         };
 
@@ -90,12 +99,21 @@ impl Codec {
             },
             #[cfg(feature = "snappy")]
             Codec::Snappy => {
-                let mut read = Vec::new();
-                {
-                    let mut reader = Reader::new(&stream[..]);
-                    reader.read_to_end(&mut read)?;
+                use byteorder::ByteOrder;
+
+                let decompressed_size = snap::decompress_len(&stream[..stream.len() - 4])?;
+                let mut decoded = vec!(0; decompressed_size);
+                snap::Decoder::new().decompress(&stream[..stream.len() - 4], &mut decoded[..])?;
+
+                let expected_crc = byteorder::BigEndian::read_u32(&stream[stream.len() - 4..]);
+                let actual_crc = crc::crc32::checksum_ieee(&decoded);
+
+                if expected_crc != actual_crc {
+                    return Err(DecodeError::new(format!("bad Snappy CRC32; expected {:x} but got {:x}",
+                                                        expected_crc,
+                                                        actual_crc)).into());
                 }
-                *stream = read;
+                *stream = decoded;
             },
         };
 

--- a/src/lib.rs
+++ b/src/lib.rs
@@ -502,6 +502,10 @@ extern crate serde;
 extern crate serde_json;
 #[cfg(feature = "snappy")]
 extern crate snap;
+#[cfg(feature = "snappy")]
+extern crate byteorder;
+#[cfg(feature = "snappy")]
+extern crate crc;
 
 // test dependency
 #[cfg(test)]

